### PR TITLE
Add a setops deployment workflow

### PR DESCRIPTION
**Pull request by hzeus**

### CODE REVIEW

File: build-and-push-image/action.yml [hunk 1]
```diff
@@ -0,0 +1,107 @@
+name: "Build and push image to SetOps registry"
+description: "Builds the docker images and pushes it to the SetOps registry while creating a tag for each stage / app combination"
```
Comment (kevinscholz) [inline build-and-push-image/action.yml:2]:
```suggestion
description: "Builds the Docker images and pushes it to the SetOps registry while creating a tag for each stage / app combination"
```

File: build-and-push-image/action.yml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+inputs:
+  setops-organization:
+    description: The SetOps organization
+    required: true
+  setops-project:
+    description: The SetOps project name
+    required: true
+  setops-username:
+    description: The SetOps username, usually obtained via secrets.SETOPS_USER
```
Comment (kevinscholz) [inline build-and-push-image/action.yml:12]: `secrets.SETOPS_USER` isn't known on this level, so I'd remove it from the description.

File: build-and-push-image/action.yml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    required: true
+  setops-password:
+    description: The SetOps password, usually obtained via secrets.SETOPS_PASSWORD
```
Comment (kevinscholz) [inline build-and-push-image/action.yml:15]: secrets.SETOPS_PASSWORD isn't known on this level, so I'd remove it from the description.

File: build-and-push-image/action.yml [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    - name: "Set up Docker Buildx"
+      uses: docker/setup-buildx-action@v1
+    - name: Cache Docker layers
+      uses: actions/cache@v2
+      with:
+        path: /tmp/.buildx-cache
+        key: ${{ runner.os }}-buildx-${{ steps.get_date.outputs.date }}-${{ inputs.build-cache-key-prefix }}-${{ github.sha }}
+        restore-keys: |
+          ${{ runner.os }}-buildx-${{ steps.get_date.outputs.date }}-${{ inputs.build-cache-key-prefix }}-
+    - name: "Install Setops"
```
Comment (kevinscholz) [inline build-and-push-image/action.yml:78]:
```suggestion
    - name: "Install SetOps"
```

File: deployment/action.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        sos changeset:discard || true
+
+        declare -A apps_with_release_id
+
+        for app in ${apps[@]}; do
+          apps_with_release_id[$app]=$(sos --app $app release:create ${{ inputs.image-digest }} | grep -o 'ReleaseID.*' | grep -o '[0-9].*')
+        done
+
+        echo "Changeset for creating releases"
```
Comment (kevinscholz) [inline deployment/action.yml:60]: I think this step could be split up in multiple steps for simplicity. Do you agree?

File: deployment/action.yml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        sos changeset:commit
+
+        predeploy_command="${{ inputs.predeploy-command }}"
```
Comment (kevinscholz) [inline deployment/action.yml:63]: Could this include double quotes? If so, should they be escaped via Github Action syntax?

**Review by kevinscholz — CHANGES_REQUESTED** [review]
Great work, this already looks really good 👏 Thanks for investing the time to provide the workflow and the actions.

I added a few questions/comments. After we addressed them I'd do a final round of testing and then we should be good to go.

File: deployment/action.yml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        for app in $(echo '${{ inputs.apps }}' | jq -r '.[]'); do
+          apps+=($app)
+        done
+
+        sos changeset:discard || true
+
+        declare -A apps_with_release_id
+
+        for app in ${apps[@]}; do
+          apps_with_release_id[$app]=$(sos --app $app release:create ${{ inputs.image-digest }} | grep -o 'ReleaseID.*' | grep -o '[0-9].*')
```
Comment (kevinscholz) [inline deployment/action.yml:57]: Linter:
```suggestion
          apps_with_release_id[$app]=$(sos --app "$app" release:create ${{ inputs.image-digest }} | grep -o 'ReleaseID.*' | grep -o '[0-9].*')
```

File: deployment/action.yml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+          first_app=${apps[0]}
+          release_id_of_first_app=${apps_with_release_id[$first_app]}
+          sos --app $first_app task:run --debug --release $release_id_of_first_app -- bash -c "${{ inputs.predeploy-command }} && echo 'SETOPS_SUCCESS'" | tee /dev/stderr | grep "SETOPS_SUCCESS" > /dev/null
+        else
+          echo "No predeploy command has been configured, skipping."
+        fi
+
+        for app in "${apps[@]}"; do
+          release_id=${apps_with_release_id[$app]}
+          sos --app $app release:activate $release_id
```
Comment (kevinscholz) [inline deployment/action.yml:75]: Linter:
```suggestion
          sos --app "$app" release:activate "$release_id"
```

File: deployment/action.yml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+        echo "Changeset for creating releases"
+        sos changeset:commit
+
+        predeploy_command="${{ inputs.predeploy-command }}"
+        if [ ${#predeploy_command} -gt 0 ]; then
+          echo "Run predeploy command"
+          first_app=${apps[0]}
+          release_id_of_first_app=${apps_with_release_id[$first_app]}
+          sos --app $first_app task:run --debug --release $release_id_of_first_app -- bash -c "${{ inputs.predeploy-command }} && echo 'SETOPS_SUCCESS'" | tee /dev/stderr | grep "SETOPS_SUCCESS" > /dev/null
```
Comment (kevinscholz) [inline deployment/action.yml:68]: Linter
```suggestion
          sos --app "$first_app" task:run --debug --release "$release_id_of_first_app "-- bash -c "${{ inputs.predeploy-command }} && echo 'SETOPS_SUCCESS'" | tee /dev/stderr | grep "SETOPS_SUCCESS" > /dev/null
```

File: deployment/action.yml (hunk 5 above)
Comment (kevinscholz) [reply deployment/action.yml:56]: Decided to not do it now

**Comment by hzeus** [comment]
@kevinscholz Thanks for the updates, looks very good to me 👍

**Pull request merged**